### PR TITLE
[RFC] kernel: Extend busy waiting with mixed mode

### LINKED ISSUE #24784: nRF: Busy wait clock is skewed vs. timer clock
On nRF5 SoCs, the integration uses CONFIG_ARCH_HAS_CUSTOM_BUSY_WAIT=y to forward k_busy_wait() calls to a HAL-defined nrfx_coredep_delay_us() function (which presumably works at CPU cycle precision) instead of using the default Zephyr one (which is limited to the 32kHz cycle rate).

Unfortunately, these clocks don't match.  The HAL timer is consistently about .5-.6% too fast, meaning that k_busy_wait() will return earlier than users expect.

This is causing a test failure on at least nrf52840 devices, where tests/kernel/timer/timer_api has a few long delays that it expects to be correct relative to the tick/cycle timer.  This got unmasked by the timer API changes (which fixed a double-round-up in the k_timer_remaining_*() paths), but it seems like the problem is definitely at the platform layer.

The code below does a bunch of busy waits at different delays and compares them to the times reported by the existing timer driver.  It produces the following log on my nrf52840dk_nrf52840.  Note that as the delays increase the error settles down to a consistent asymptote, implying that it's the clock rates that are skewed and not just an offset or aliased conversion somewhere:

(Note also that as I iterate the test, the deltas are jumping around quite a bit to the tune of 5-10 32kHz cycles per half-second busy wait.  300us of jitter seems like **way** too much to me given the clock rates involved.)

```
*** Booting Zephyr OS build zephyr-v2.2.0-2044-g315cba1c40fc  ***
                eff   err
   us    cyc     us   ppm
 ===== ====== ====== =====
     1      0      0 1000000
     2      0      0 1000000
     4      0      0 1000000
     8      0      0 1000000
    16      1     30     0
    32      1     30 500000
    64      2     61 333333
   128      5    152     0
   256      8    244 111111
   512     16    488 58823
  1024     33   1007 29411
  2048     67   2044 14705
  4096    134   4089  7407
  8192    267   8148  7434
 16384    535  16326  3724
 32768   1070  32653  3724
 65536   2141  65338  3258
131072   4275 130462  4656
262144   8542 260681  5587
524288  17080 521240  5820
```

```c
#include <zephyr.h>
#include <stdlib.h>

void main(void)
{
	printk("                eff   err\n");
	printk("   us    cyc     us   ppm\n");
	printk(" ===== ====== ====== =====\n");
	for (int us = 1; us < 1000000; us *= 2) {

		/* Busy Wait */
		int t0 = k_cycle_get_32();
		k_busy_wait(us);
		int t1 = k_cycle_get_32();

		/* Effective us waited: floor, because we want to wait
		 * at least this long and this is the longest duration
		 * we can prove.
		 */
		int eff_us = k_cyc_to_us_floor32(t1 - t0);

		/* Error, in parts per million */
		int exp_cyc = k_us_to_cyc_ceil32(us);
		int err_ppm = (int)(1000000ull
				    * abs((t1 - t0) - exp_cyc) / exp_cyc);

		printk("%6d %6d %6d %5d\n", us, t1-t0, eff_us, err_ppm);
	}
}
```

### PR DESCRIPTION
**Pull request by nordic-krch**
If custom busy wait is used then it is possible that it has different
clock source than system clock and time measured will be different
compared to one sourced from system clock due to clock sources
inaccuracy. On the other hand custom busy waiting is used when system
clock frequency is low and cannot handle delays with microseconds
precision.

Adding CONFIG_ARCH_HAS_CUSTOM_MIXED_BUSY_WAIT option which is using
system clock for coarse busy waiting and custom function for
remainder. With this approach, error that could be seen for longer
delays is reduced.

Fixed #24784.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

**Review by pabigot — CHANGES_REQUESTED** [review]
Without looking at the implementation this needs a lot of work.

My standard test for this involves a monitoring a scope pin with a Logic-Pro as it bounds `k_busy_wait()` calls for these durations in microseconds:

		1, 2, 5, 10, 15, 20,
		50, 50, 50,
		75, 75, 75,
		100, 100, 100,
		1000000,

Using current master the measured durations for two test boards are:

		 * particle_xenon
		 * 2.22, 2.94, 5.92, 10.92, 15.88, 20.86,
		 * 50.78, 50.76, 50.76
		 * 75.68, 75.66, 75.66
		 * 100.6, 100.6, 100.6
		 * 996.9 (ms)
		 *
		 * pca10040
		 * 2.22, 3.12, 6.12, 11.12, 16.12, 21.16
		 * 51.18, 51.18, 51.2
		 * 76.2, 76.26, 76.24,
		 * 101.3, 101.2, 101.3
		 * 1001 (ms)


These are acceptable results: the sleep is reasonably close to the requested duration and biased to be longer rather than shorter.

With this PR I got these results:

		 * particle_xenon
		 * 3.34, 2.96, 2.96, 2.96, 2.96, 3,
		 * 50.8, 26.5, 27.3,
		 * 51.32, 57.82, 56.98,
		 * 82.78, 88.44, 87.64
		 * 1001 (ms)
		 *
		 * pca10040
		 * does not start

pca10040 also doesn't run hello_world for me in this branch, but it's the particle_xenon results that are the concern.

The effective delay with this PR for durations between 5 us and 25 us are unacceptably low.  The same for 75 us and 100 us.  The behavior for a delay of 50 us is off (in the wrong direction) 2 out of 3 samples.

The PR does do the "right" thing for the atypical 1 s delay.

I'll clean up the sample and make it available if you want to pursue this further.

File: kernel/Kconfig [hunk 1]
```diff
@@ -521,6 +521,18 @@ config ARCH_HAS_CUSTOM_BUSY_WAIT
 	  the provided k_busy_wait(), but instead must do something custom. It must
 	  enable this option in that case.
 
+if ARCH_HAS_CUSTOM_BUSY_WAIT
+config ARCH_HAS_CUSTOM_MIXED_BUSY_WAIT
```
Comment (andyross) [inline kernel/Kconfig:525]: This isn't an ARCH feature, it's not even board or soc-specific.  It has to do with whether or not the clock used to drive a custom busy wait is stable and calibrated or not, which is an application-level configuration choice even on nRF5.

I think I'd prefer this be handled at the platform layer, which is the point of "custom busy wait" to begin with.  It's not clear to me that there will be much benefit to having complicated generic code for this in the (currently very simple) k_busy_wait() layer.

**Comment by nordic-krch** [comment]
@pabigot yes, i see that due to calculation time and alignment of RTC ticks "mixed" method will never give good results for us range. It starts to show better results for delays > 10ms.

@andyross I don't agree that this is nordic specific case. It might be like that now but zephyr is targeting low power devices and it will be applicable to any device that has low frequency system clock and thus different clock source for cpu and system clock. Even if HF clock is XTAL you might have LF clock from RC.

Mixed mode won't serve. What about threshold based one? Custom busy wait is used only for values below configurable threshold.

**Comment by pabigot** [comment]
> Mixed mode won't serve. What about threshold based one? Custom busy wait is used only for values below configurable threshold.

Only if you can guarantee that `k_busy_wait(x)` and `k_busy_wait(x+1)` have a consistent relationship for all x.  Having a significant discontinuity, or non-monotonic behavior, would be a problem.

While Zephyr does target low-power devices it's a long way from supporting them fully, and it will never be the only target application.

At this time I'd prefer a mix of two solutions to #24784:
* Default to precise/accurate clocks for both CPU and system timer, allowing people who care more about power than accuracy to opt in to an alternative;
* Accept that the system clock may be different from the CPU clock and that therefore `k_busy_wait()` may produce a different delay than  `k_sleep()` for the same requested duration.

That position may change if it's clear that a significant number of cross-vendor boards select unstable clocks by default, but the discussion for #22758 suggests the consensus approach is to default stable.

**Comment by nordic-krch** [comment]
> Only if you can guarantee that k_busy_wait(x) and k_busy_wait(x+1) 

that cannot be guaranteed.

> Default to precise/accurate clocks for both CPU and system timer, allowing people who care more about power than accuracy to opt in to an alternative;

That is only possible on boards with LF XTAL.

>  but the discussion for #22758 suggests the consensus approach is to default stable.

I think that there is a different meaning of stable there. Here we are talking about accuracy and HF clock accuracy is not that important except for k_busy_wait().

So it seems that there is no solution that could be accepted by all of us. For test purposes we can use   `CLOCK_CONTROL_NRF_K32SRC_SYNTH` assuming that k_busy_wait() is used only for short periods (<10 ms range) in the application.

**Comment by nordic-krch** [comment]
Created alternative which should have clocks aligned for testing: #24972

**Comment by pabigot** [comment]
> > Only if you can guarantee that k_busy_wait(x) and k_busy_wait(x+1)
> 
> that cannot be guaranteed

unless it uses the same clock for all durations, and that clock is stable, I agree.  We can fix the stability part at least where there's a policy to wait for stabilization before proceeding to boot.

> I think that there is a different meaning of stable there. Here we are talking about accuracy and HF clock accuracy is not that important except for k_busy_wait().

`k_busy_wait()` is pretty important, since drivers depend on it to meet datasheet-specified durations.  Those should have some tolerance in them, so 8% maximum absolute error from HFINT might be OK.

> So it seems that there is no solution that could be accepted by all of us. For test purposes we can use `CLOCK_CONTROL_NRF_K32SRC_SYNTH` assuming that k_busy_wait() is used only for short periods (<10 ms range) in the application.

My results checking things with an external 1pps signal on one nrf52840dk_nrf52840 device were:

    When HFCLK is on the timer and systick errors are less than 1 ppm.  In
    that case synth is the highest accuracy LFCLK source, with XTAL then
    RC in decreasing accuracy.
    
    When HFCLK is off the timer and systick errors are the same and are
    generally between 3000 and 4000 ppm.  Using the synth clock LFCLK
    tracks the HFCLK error; RC reduces to between -30 and -60 ppm (30 ppm
    resolution); XTAL between 0 and 30 ppm (longer samples show 25 ppm).

So if what's important is minimizing the difference between elapsed time using `k_busy_wait()` and `k_sleep()` then switching to synth as a default would be the best option regardless of whether there's a crystal.

**Pull request closed without merging**